flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -114,13 +128,33 @@
 
     private final StateChangelogWriter<?> stateChangelogWriter;
 
+    private long lastCheckpointId = -1L;
+
     /** last accessed partitioned state. */
     @SuppressWarnings("rawtypes")
     private InternalKvState lastState;
 
     /** For caching the last accessed partitioned state. */
     private String lastName;
 
+    /** Updated initially on restore and later upon materialization (after FLINK-21356). */
+    private final List<KeyedStateHandle> materialized = new ArrayList<>();
```

Comment:
this should be changeable right?
remove final?

Reply:
Right (if updated by changing reference it shouldn't be `final` as opposed to `clear` + `addAll`).